### PR TITLE
implement `drop view`

### LINKED ISSUE #3251: implement `DROP VIEW`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

**Describe the solution you'd like**

Views can only be dropped with `drop table` currently. Should support `drop view` syntax and change `drop table` so it only works against tables, not views.

**Describe alternatives you've considered**

**Additional context**

for completeness it should be implemented, but also noticed it's used in q15 of TPC-H

### PR DESCRIPTION
**Pull request by kmitchener**
# Which issue does this PR close?

Closes #3251.

 # Rationale for this change

# What changes are included in this PR?

* implement `drop view` and `drop view if exists`
* change `drop table` so it only works on tables, not views
* added tests for the above
* added basic documentation for `drop view`

# Are there any user-facing changes?

### CODE REVIEW

**Comment by avantgardnerio** [comment]
@DaltonModlin may be interested in this.

**Comment by kmitchener** [comment]
Functionality plus basic tests and docs added

**Comment by kmitchener** [comment]
One thing I didn't do in this PR: re-export the new datafusion_expr::logical_plan::DropView struct via the "legacy" logical_plan module in core (datafusion::logical_plan:: ..). Should I re-export it or does legacy in this case mean that no new re-exports should happen there?

https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/logical_plan/mod.rs#L18

Alternatively, should we create an issue to remove this legacy re-exporting?

**Comment by andygrove** [comment]
> Alternatively, should we create an issue to remove this legacy re-exporting?

We do have an issue already for this - https://github.com/apache/arrow-datafusion/issues/2683

We may want to deprecate this module for one release and then remove in the next.

File: docs/source/user-guide/sql/ddl.md [hunk 1]
```diff
@@ -94,14 +94,28 @@ CREATE TABLE memtable as select * from valuetable;
...
 
 ```sql
 CREATE TABLE users AS VALUES(1,2),(2,3);
-
 DROP TABLE users;
+-- or use 'if exists' to silently ignore if the table doesn't exist
+DROP TABLE IF EXISTS nonexistent_table;
+```
+
+## DROP VIEW
```
Comment (andygrove) [inline docs/source/user-guide/sql/ddl.md:110]: Thanks for improving the docs!

File: docs/source/user-guide/sql/ddl.md [hunk 2]
```diff
@@ -94,14 +94,28 @@ CREATE TABLE memtable as select * from valuetable;
...
+
+Removes the view from DataFusion's catalog.add
```
Comment (andygrove) [inline docs/source/user-guide/sql/ddl.md:112]: Looks like a typo at the end of this line?

Comment (kmitchener) [reply docs/source/user-guide/sql/ddl.md]: yep. fixed, thanks.

**Comment by kmitchener** [comment]
@andygrove This is ready. Let me know what I can do to help get it committed. Thanks!

**Comment by kmitchener** [comment]
@alamb is there someone that's suitable to review this and give feedback? this PR is blocking progress on other issues and it's not getting a review :( and I'm not sure how to proceed

**Comment by alamb** [comment]
SOrry @kmitchener  -- I will review it now

File: datafusion/sql/src/planner.rs [hunk 3]
```diff
@@ -245,20 +245,29 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                }))
-            }
+                })),
+                ObjectType::View => Ok(LogicalPlan::DropView(DropView {
+                    name: names.get(0).unwrap().to_string(),
+                    if_exists,
+                    schema: DFSchemaRef::new(DFSchema::empty()),
+                })),
+                _ => Err(DataFusionError::NotImplemented(
+                    "Only `DROP TABLE/VIEW  ...` statement is supported currently"
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:267]: 👍

File: datafusion/expr/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -1019,6 +1030,17 @@ pub struct DropTable {
     pub schema: DFSchemaRef,
 }
 
+/// Drops a view.
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1033]: Not for this PR, but it seems to me like requiring new `LogicalPlan` nodes for each type of  DDL (`CREATE`, `DROP`, etc) is somewhat cumbersome. Maybe we can eventually refactor type of thing into `LogicalPlan::DDL(DDL)` and keep all the DDL related structs in their own structure. I'll file a follow on ticket for this

File: datafusion/core/src/execution/context.rs [hunk 5]
```diff
@@ -313,10 +310,7 @@ impl SessionContext {
                 let table = self.table(name.as_str());
 
                 match (if_not_exists, or_replace, table) {
-                    (true, false, Ok(_)) => {
-                        let plan = LogicalPlanBuilder::empty(false).build()?;
-                        Ok(Arc::new(DataFrame::new(self.state.clone(), &plan)))
-                    }
+                    (true, false, Ok(_)) => self.return_empty_dataframe(),
```
Comment (alamb) [inline datafusion/core/src/execution/context.rs:313]: This change to refactor and reduce duplicated code, is a very nice example of leaving the code better than when you found it. It is very much appreciated @kmitchener

File: datafusion/core/tests/sql/create_drop.rs [hunk 6]
```diff
@@ -112,6 +112,104 @@ async fn drop_table() -> Result<()> {
...
+        .await
+        .unwrap()
+        .collect()
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+#[should_panic(expected = "doesn't exist")]
+async fn drop_view_cant_drop_table() {
```
Comment (alamb) [inline datafusion/core/tests/sql/create_drop.rs:151]: 👍 nice

**Review by alamb — APPROVED** [review]
This looks great @kmitchener  -- thank you and I apologize again for the delay in review

File: datafusion/expr/src/logical_plan/plan.rs (hunk 4 above)
Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:1033]: Filed https://github.com/apache/arrow-datafusion/issues/3349 as a follow on

Comment (kmitchener) [reply datafusion/expr/src/logical_plan/plan.rs:1033]: Yes, I was wondering about this and follow on work for dropping schema, etc. A lot of plumbing here, and I'm not sure it works if you have your own CatalogProvider.

Comment (alamb) [reply datafusion/expr/src/logical_plan/plan.rs:1033]: > Yes, I was wondering about this and follow on work for dropping schema, etc. A lot of plumbing here, and I'm not sure it works if you have your own CatalogProvider.

I agree -- it does not work if you have your own CatalogProvider without some more work. I this behavior is fine, as long as the point is clear to users.

**Pull request merged**